File: Java/0079_word_search/0079_word_search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        if (board.length == 0 || board[0].length == 0)  
            return false;  
        int m = board.length, n = board[0].length;    
        boolean[][] visited = new boolean[m][n];    
        for (int i = 0; i < m; i++)    
            for (int j = 0; j < n; j++)    
                if (dfs(board, visited, word, i, j, 0))   
                    return true;    
        return false; 
    }
    
    public boolean dfs(char[][] board, boolean[][] visited, String word, int x, int y, int k) {  
        if (k >= word.length())  
            return true;  
        int m = board.length, n = board[0].length;  
        if (x < 0 || x >= m || y < 0 || y >= n || visited[x][y] || board[x][y] != word.charAt(k))  
            return false;  
        visited[x][y] = true;  
        if (dfs(board, visited, word, x + 1, y, k + 1))  
            return true;  
        if (dfs(board, visited, word, x - 1, y, k + 1))  
            return true;  
        if (dfs(board, visited, word, x, y + 1, k + 1))  
            return true;  
        if (dfs(board, visited, word, x, y - 1, k + 1))  
            return true;  
        visited[x][y] = false;  
        return false;
    }
}

File: Python/0079_word_search/0079_word_search.py
class Solution:
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        if len(board) == 0 or len(board[0]) == 0:
            return False
        m = len(board)
        n = len(board[0])
        visited = [[False for i in range(n)] for j in range(m)]
        for i in range(m):
            for j in range(n):
                if self.dfs(board, visited, word, i, j, 0):
                    return True
        return False
              
    def dfs(self, board, visited, word, x, y, k):
        if k >= len(word):
            return True
        m = len(board)
        n = len(board[0])
        if x < 0 or x >= m or y < 0 or y >= n or visited[x][y] or board[x][y] != word[k]:
            return False
        visited[x][y] = True
        if self.dfs(board, visited, word, x + 1, y, k + 1):
            return True
        if self.dfs(board, visited, word, x - 1, y, k + 1):
            return True
        if self.dfs(board, visited, word, x, y + 1, k + 1):
            return True
        if self.dfs(board, visited, word, x, y - 1, k + 1):
            return True
        visited[x][y] = False
        return False
 